{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running a WF simulation in `shadie`\n",
    "\n",
    "We will start with a simple example using the Wright-Fisher model.\n",
    "\n",
    "`shadie` provides a standard Wright-Fisher Model, as well as some alternatives (e.g. the Moran model). These models are implemented as SLiM nonWF models. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running the Wright-Fisher\n",
    "The classic Wright-Fisher Model makes the following assumptions: \n",
    "1. Discrete, non-overlapping generations\n",
    "2. Random mating between hermaphroditic individuals\n",
    "3. Constant population size (of diploid individuals; 2N haplotypes)\n",
    "4. No fitness effects; simulation is neutral (no selection)\n",
    "5. No recombination\n",
    "    - Note that the default recombination rate in `shadie` is NOT 0.0, so to meet this requirement you must be sure to set it in the initialize function\n",
    "6. Infinite sites\n",
    "\n",
    "`shadie` uses context manager to check model parameters before running a simulation. The syntax requires the user to define a model using a `with` statement and then defining all parameters. The model needs a `.initialize()` function, which takes general parameters such as the chromosome, mutation rate, recombination rate, and file names. You can also specify what kind of reproduction the model will use with `.reproduction.` followed by the appropriate function (a list of options can be seen in Jupyter Notebooks by pressing tab after typing `.reproduction.`.\n",
    "\n",
    "See the example below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shadie\n",
    "default_chrom = shadie.chromosome.default()\n",
    "\n",
    "with shadie.Model() as WF_model:\n",
    "    WF_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        recomb_rate=0.0,\n",
    "        sim_time=1000,\n",
    "        file_out=\"WF.trees\",\n",
    "    )\n",
    "    WF_model.reproduction.wright_fisher(pop_size=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, your model exists as the model object you defined, in this case `WF_model`. You can print the model script by running:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "initialize() {\n",
      "\n",
      "  // model type\n",
      "  initializeSLiMModelType(\"nonWF\");\n",
      "\n",
      "  // config\n",
      "  initializeRecombinationRate(0.0, 10000);\n",
      "  initializeMutationRate(1e-08);\n",
      "  initializeTreeSeq(simplificationInterval=NULL);\n",
      "\n",
      "  // MutationType init\n",
      "  initializeMutationType('m2', 0.1, 'g', -3.0, 1.5);\n",
      "  initializeMutationType('m3', 0.8, 'e', 0.04);\n",
      "  c(m3, m2).haploidDominanceCoeff = 1.0;\n",
      "  c(m3, m2).convertToSubstitution = T;\n",
      "\n",
      "  \n",
      "\n",
      "  // ElementType init\n",
      "  initializeGenomicElementType('g1', c(m2,m3), c(8,0.1));\n",
      "  initializeGenomicElementType('g2', m2, 1);\n",
      "\n",
      "  // Chromosome (GenomicElement init)\n",
      "  types = rep(g1, asInteger(floor(1999/3)));\n",
      "  starts = 2001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "  types = rep(g2, asInteger(floor(1999/3)));\n",
      "  starts = 4001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "  types = rep(g1, asInteger(floor(1999/3)));\n",
      "  starts = 6001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "\n",
      "  // constants (Population sizes and others)\n",
      "  defineConstant('K', 1000);\n",
      "\n",
      "  // globals (metadata dictionary)\n",
      "  defineGlobal('METADATA', Dictionary('file_in', 'None', 'file_out', 'WF.trees', 'mutation_rate', '1e-08', 'recomb_rate', '0.0', 'model', 'shadie WF', 'length', '1000', 'spo_pop_size', '1000', 'gam_pop_size', 'NA', 'spo_mutation_rate', '1e-08', 'recombination_rate', '0.0', 'selection', 'none', 'gens_per_lifecycle', '1'));\n",
      "\n",
      "  // extra scripts (Optional)\n",
      "  \n",
      "}\n",
      "\n",
      "\n",
      "// executes before offspring are generated\n",
      "// define starting diploid population.\n",
      "\n",
      "1 first() {\n",
      "    sim.addSubpop('p1', K);\n",
      "}\n",
      "\n",
      "\n",
      "// generates offspring\n",
      "// WF model with no selection; random mating\n",
      "reproduction(p1 ) {\n",
      "    subpop.addCrossed(individual, subpop.sampleIndividuals(1));\n",
      "}\n",
      "\n",
      "\n",
      "// implements survival adjustments\n",
      "// non-overlapping generations\n",
      "survival() {\n",
      "    return (individual.age == 0);\n",
      "}\n",
      "\n",
      "\n",
      "// executes after selection occurs\n",
      "// end of sim; save .trees file\n",
      "\n",
      "1001 late() {\n",
      "    sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);\n",
      "    sim.treeSeqOutput('WF.trees', metadata = METADATA);\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(WF_model.script)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If desired, this script can be copied and pasted into the SLiMgui, or saved as a file and run using the command line. \n",
    "\n",
    "To run the model in `shadie`, simply call the `run()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "WF_model.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output of the SLiM simulations, a `.trees` file with the indicated `file_out` name, should now be located in your working directory.\n",
    "\n",
    "### Access the Tree Sequence\n",
    "\n",
    "You can inspect the output of the simulation using the `postsim` module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸŒ¿ INFO | \u001b[1mone_sim.py\u001b[0m | \u001b[30mshadie assumes sims were run without a burn-in and without neutral mutations and will recapitate and mutate (add neutral mutations) any loaded sims by default. Current settings:\n",
      "Recapitate: True\n",
      "Add neutral mutations: True\n",
      "\u001b[0m\n",
      "ðŸŒ¿ INFO | \u001b[1mone_sim.py\u001b[0m | \u001b[30mSimulation currently has 1 existing mutations of type(s) {2}.\u001b[0m\n",
      "ðŸŒ¿ INFO | \u001b[1mone_sim.py\u001b[0m | \u001b[30mMutating using mutation rate: 1e-08. Keeping 6 existing mutations of type(s) {0, 2}.\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>10001.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>ticks</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>7513</td></tr>\n",
       "                      <tr><td>Total Size</td><td>783.9 KiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 1001<br/>file_version: 0.8<br/>model_type: nonWF<br/>name: sim<br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: late<br/>tick: 1001<br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">user_metadata:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">file_in:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     None<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">file_out:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     WF.trees<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">gam_pop_size:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     NA<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">gens_per_lifecycle:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     1<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">length:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     1000<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">model:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     shadie WF<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">mutation_rate:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     1e-08<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">recomb_rate:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     0.0<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">recombination_rate:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     0.0<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">selection:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     none<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">spo_mutation_rate:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     1e-08<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">spo_pop_size:</span>\n",
       "                  <details >\n",
       "                    <summary>list</summary>\n",
       "                     1000<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>7520</td>\n",
       "                      <td>235.0 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>2000</td>\n",
       "                      <td>197.1 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>6</td>\n",
       "                      <td>1.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>7521</td>\n",
       "                      <td>279.7 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>2</td>\n",
       "                      <td>2.4 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        âœ…\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>4</td>\n",
       "                      <td>7.0 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>6</td>\n",
       "                      <td>160 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7ff759ef25e0>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "postsim = shadie.postsim.OneSim(\"WF.trees\", default_chrom)\n",
    "postsim.tree_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the tree sequence using the `draw_tree_sequence()` function. This function samples 10 haplotypes by default:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"tf61d0355547449b39a655c17631fed6e\" style=\"text-align:center;width:300.0px\"><svg class=\"toyplot-canvas-Canvas\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns=\"http://www.w3.org/2000/svg\" width=\"300.0px\" height=\"425.0px\" viewBox=\"0 0 300.0 425.0\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" id=\"teca4e46dea6c433b84a67bab9b740d9e\"><g class=\"toyplot-coordinates-Cartesian\" id=\"t3ffb2b0110ec44eb889deed050b5f2c3\"><clipPath id=\"t75581f6a57d14507b902d5970467361f\"><rect x=\"45.0\" y=\"45.0\" width=\"210.0\" height=\"35.0\"></rect></clipPath><g clip-path=\"url(#t75581f6a57d14507b902d5970467361f)\"><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"t14d72f9fec834daf987af88a098a04e3\"><polygon points=\"50.0,75.0 90.0,75.0 90.0,50.0 50.0,50.0\" style=\"fill:rgb(16.9%,51.4%,72.9%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: noncds\n",
       "interval: (0, 2000)\n",
       "ElementType: g3\n",
       "is coding: False</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"t93f3ceba288a42638a3724daead8545b\"><polygon points=\"90.02000000000001,75.0 130.0,75.0 130.0,50.0 90.02000000000001,50.0\" style=\"fill:rgb(67.1%,86.7%,64.3%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: exon\n",
       "interval: (2001, 4000)\n",
       "ElementType: g1\n",
       "is coding: True</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"tbabcb03ce25d463299d6f2a6189b6e8b\"><polygon points=\"130.02,75.0 170.0,75.0 170.0,50.0 130.02,50.0\" style=\"fill:rgb(99.2%,68.2%,38%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: intron\n",
       "interval: (4001, 6000)\n",
       "ElementType: g2\n",
       "is coding: True</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"ta2b528c2ef56434793d4f40203a283fe\"><polygon points=\"170.02,75.0 210.0,75.0 210.0,50.0 170.02,50.0\" style=\"fill:rgb(67.1%,86.7%,64.3%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: exon\n",
       "interval: (6001, 8000)\n",
       "ElementType: g1\n",
       "is coding: True</title></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" id=\"tbd21d96938ce49939647154b2e7665d2\"><polygon points=\"210.02,75.0 250.0,75.0 250.0,50.0 210.02,50.0\" style=\"fill:rgb(16.9%,51.4%,72.9%);fill-opacity:1.0;opacity:0.75;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\"><title>name: noncds\n",
       "interval: (8001, 10000)\n",
       "ElementType: g3\n",
       "is coding: False</title></polygon></g></g><g class=\"toyplot-coordinates-Axis\" id=\"t76d36694ce7f432eb5465e0f6a8f47f4\" transform=\"translate(50.0,75.0)translate(0,5.0)\"><line x1=\"0\" y1=\"0\" x2=\"200.0\" y2=\"0\" style=\"\"></line><g><line x1=\"0.0\" y1=\"5.0\" x2=\"0.0\" y2=\"0\" style=\"\"></line><line x1=\"40.0\" y1=\"5.0\" x2=\"40.0\" y2=\"0\" style=\"\"></line><line x1=\"80.0\" y1=\"5.0\" x2=\"80.0\" y2=\"0\" style=\"\"></line><line x1=\"120.0\" y1=\"5.0\" x2=\"120.0\" y2=\"0\" style=\"\"></line><line x1=\"160.0\" y1=\"5.0\" x2=\"160.0\" y2=\"0\" style=\"\"></line><line x1=\"200.0\" y1=\"5.0\" x2=\"200.0\" y2=\"0\" style=\"\"></line></g><g><g transform=\"translate(0.0,6)\"><text x=\"-2.78\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">0</text></g><g transform=\"translate(40.0,6)\"><text x=\"-11.12\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">2000</text></g><g transform=\"translate(80.0,6)\"><text x=\"-11.12\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">4000</text></g><g transform=\"translate(120.0,6)\"><text x=\"-11.12\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">6000</text></g><g transform=\"translate(160.0,6)\"><text x=\"-11.12\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">8000</text></g><g transform=\"translate(200.0,6)\"><text x=\"-13.9\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">10000</text></g></g><g class=\"toyplot-coordinates-Axis-coordinates\" style=\"visibility:hidden\" transform=\"\"><line x1=\"0\" x2=\"0\" y1=\"-3.0\" y2=\"4.5\" style=\"stroke:rgb(43.9%,50.2%,56.5%);stroke-opacity:1.0;stroke-width:1.0\"></line><text x=\"0\" y=\"-6\" style=\"alignment-baseline:alphabetic;fill:rgb(43.9%,50.2%,56.5%);fill-opacity:1.0;font-size:10px;font-weight:normal;stroke:none;text-anchor:middle\"></text></g></g></g><g class=\"toyplot-coordinates-Cartesian\" id=\"t2a63a98fea8845d1a9f80d3e10601e03\"><clipPath id=\"t3ccf4bd86b0a4372bce8efe01564286a\"><rect x=\"40.0\" y=\"90.0\" width=\"220.0\" height=\"295.0\"></rect></clipPath><g clip-path=\"url(#t3ccf4bd86b0a4372bce8efe01564286a)\"><g class=\"toyplot-mark-FillBoundaries\" style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:0.85;stroke:rgb(100%,100%,100%);stroke-opacity:0.5\" id=\"ta97350ec521748418ca035b28844511b\"><polygon points=\"50.0,108.76760767491741 250.0,108.76760767491741 250.0,100.0 50.0,100.0\" style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:0.85;opacity:1.0;stroke:rgb(100%,100%,100%);stroke-opacity:0.5\"></polygon></g><g class=\"toyplot-mark-FillBoundaries\" style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:0.25;stroke:rgb(100%,100%,100%);stroke-opacity:0.5\" id=\"t00aede1bbebf45ed86311f7e42a02423\"><polygon points=\"50.0,108.76760767491741 65.38461538461539,143.838038374587 65.38461538461539,319.19019187293486 234.6153846153846,319.19019187293486 234.6153846153846,143.838038374587 250.0,108.76760767491741 250.0,108.76760767491741 234.6153846153846,108.76760767491741 234.6153846153846,108.76760767491741 65.38461538461539,108.76760767491741 65.38461538461539,108.76760767491741 50.0,108.76760767491741\" style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:0.25;opacity:1.0;stroke:rgb(100%,100%,100%);stroke-opacity:0.5\"><title>idx: 0\n",
       "interval: (0 - 10001)\n",
       "tmrca: 5741.62</title></polygon></g><g class=\"toytree-mark-Toytree\" id=\"t3cd0a9d52ecb468fb7439f69ff6467a7\"><g class=\"toytree-Edges\" style=\"stroke:rgb(14.5%,14.5%,14.5%);stroke-opacity:1.0;stroke-linecap:round;stroke-width:2.0;fill:none\"><path d=\"M 92.3 307.7 L 80.8 307.7 L 80.8 319.2\" id=\"11,0\" style=\"\"></path><path d=\"M 103.8 313.1 L 96.2 313.1 L 96.2 319.2\" id=\"10,1\" style=\"\"></path><path d=\"M 103.8 313.1 L 111.5 313.1 L 111.5 319.2\" id=\"10,2\" style=\"\"></path><path d=\"M 138.5 294.4 L 126.9 294.4 L 126.9 319.2\" id=\"13,3\" style=\"\"></path><path d=\"M 150.0 318.2 L 142.3 318.2 L 142.3 319.2\" id=\"12,4\" style=\"\"></path><path d=\"M 150.0 318.2 L 157.7 318.2 L 157.7 319.2\" id=\"12,5\" style=\"\"></path><path d=\"M 180.8 311.0 L 173.1 311.0 L 173.1 319.2\" id=\"14,6\" style=\"\"></path><path d=\"M 180.8 311.0 L 188.5 311.0 L 188.5 319.2\" id=\"14,7\" style=\"\"></path><path d=\"M 211.5 296.9 L 203.8 296.9 L 203.8 319.2\" id=\"15,8\" style=\"\"></path><path d=\"M 211.5 296.9 L 219.2 296.9 L 219.2 319.2\" id=\"15,9\" style=\"\"></path><path d=\"M 92.3 307.7 L 103.8 307.7 L 103.8 313.1\" id=\"11,10\" style=\"\"></path><path d=\"M 129.8 143.8 L 92.3 143.8 L 92.3 307.7\" id=\"18,11\" style=\"\"></path><path d=\"M 138.5 294.4 L 150.0 294.4 L 150.0 318.2\" id=\"13,12\" style=\"\"></path><path d=\"M 167.3 265.8 L 138.5 265.8 L 138.5 294.4\" id=\"17,13\" style=\"\"></path><path d=\"M 196.2 279.8 L 180.8 279.8 L 180.8 311.0\" id=\"16,14\" style=\"\"></path><path d=\"M 196.2 279.8 L 211.5 279.8 L 211.5 296.9\" id=\"16,15\" style=\"\"></path><path d=\"M 167.3 265.8 L 196.2 265.8 L 196.2 279.8\" id=\"17,16\" style=\"\"></path><path d=\"M 129.8 143.8 L 167.3 143.8 L 167.3 265.8\" id=\"18,17\" style=\"\"></path></g><g class=\"toytree-AdmixEdges\" style=\"fill:rgb(0.0%,0.0%,0.0%);fill-opacity:0.0;stroke:rgb(90.6%,54.1%,76.5%);stroke-opacity:0.6;font-size:14px;stroke-linecap:round;stroke-width:5\"></g><g class=\"toytree-Nodes\" style=\"fill:rgb(40.0%,76.1%,64.7%);fill-opacity:1.0;stroke:rgb(14.9%,14.9%,14.9%);stroke-opacity:1.0;stroke-width:1.5\"></g><g class=\"toytree-TipLabels\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0;font-family:Helvetica;font-size:10px;font-weight:300;vertical-align:baseline;white-space:pre;stroke:none\"><g class=\"toytree-TipLabel\" transform=\"translate(80.7692,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-4</text></g><g class=\"toytree-TipLabel\" transform=\"translate(96.1538,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-5</text></g><g class=\"toytree-TipLabel\" transform=\"translate(111.538,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-7</text></g><g class=\"toytree-TipLabel\" transform=\"translate(126.923,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-2</text></g><g class=\"toytree-TipLabel\" transform=\"translate(142.308,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-1</text></g><g class=\"toytree-TipLabel\" transform=\"translate(157.692,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-3</text></g><g class=\"toytree-TipLabel\" transform=\"translate(173.077,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-0</text></g><g class=\"toytree-TipLabel\" transform=\"translate(188.462,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-9</text></g><g class=\"toytree-TipLabel\" transform=\"translate(203.846,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-6</text></g><g class=\"toytree-TipLabel\" transform=\"translate(219.231,319.19)rotate(90)\"><text x=\"15.0\" y=\"2.5549999999999997\" style=\"fill:rgb(14.5%,14.5%,14.5%);fill-opacity:1.0\">p0-8</text></g></g></g><g class=\"toyplot-mark-Point\" id=\"t3ededf73bb534bd895d15f8d8f4904a8\"><g class=\"toyplot-Series\"><g style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:1.0;opacity:1.0;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" class=\"toyplot-Datum\" transform=\"translate(92.3076923076923, 209.3329390053765)\"><title>id: 0\n",
       "pos: 0\n",
       "time: 3597\n",
       "mtype: 0</title><circle r=\"3.5\"></circle></g><g style=\"fill:rgb(40%,76.1%,64.7%);fill-opacity:1.0;opacity:1.0;stroke:rgb(0%,0%,0%);stroke-opacity:1.0\" class=\"toyplot-Datum\" transform=\"translate(167.30769230769232, 243.0824971421263)\"><title>id: 1\n",
       "pos: 1\n",
       "time: 2492\n",
       "mtype: 0</title><circle r=\"3.5\"></circle></g></g></g><g class=\"toyplot-mark-AxisLines\" style=\"stroke-dasharray:4,2;stroke-opacity:0.4\" id=\"td068d996e60b4f50b020369db21e2d9f\"><g class=\"toyplot-Series\"><line class=\"toyplot-Datum\" y1=\"288.61909064943603\" y2=\"288.61909064943603\" x1=\"50.0\" x2=\"250.0\" style=\"opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-dasharray:4,2;stroke-opacity:0.4\"></line></g></g></g><g class=\"toyplot-coordinates-Axis\" id=\"tfdc3c3b2c16147eaaedbd49cb9fcb095\" transform=\"translate(50.0,375.0)rotate(-90.0)translate(0,-10.0)\"><line x1=\"55.80980812706508\" y1=\"0\" x2=\"275.0\" y2=\"0\" style=\"\"></line><g><line x1=\"55.80980812706508\" y1=\"0\" x2=\"55.80980812706508\" y2=\"5\" style=\"\"></line><line x1=\"116.89092945273713\" y1=\"0\" x2=\"116.89092945273713\" y2=\"5\" style=\"\"></line><line x1=\"177.97205077840917\" y1=\"0\" x2=\"177.97205077840917\" y2=\"5\" style=\"\"></line><line x1=\"239.0531721040812\" y1=\"0\" x2=\"239.0531721040812\" y2=\"5\" style=\"\"></line></g><g><g transform=\"translate(55.80980812706508,-6)\"><text x=\"-2.78\" y=\"-4.440892098500626e-16\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">0</text></g><g transform=\"translate(116.89092945273713,-6)\"><text x=\"-11.12\" y=\"-4.440892098500626e-16\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">2000</text></g><g transform=\"translate(177.97205077840917,-6)\"><text x=\"-11.12\" y=\"-4.440892098500626e-16\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">4000</text></g><g transform=\"translate(239.0531721040812,-6)\"><text x=\"-11.12\" y=\"-4.440892098500626e-16\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">6000</text></g></g><g class=\"toyplot-coordinates-Axis-coordinates\" style=\"visibility:hidden\" transform=\"\"><line x1=\"0\" x2=\"0\" y1=\"3.0\" y2=\"-4.5\" style=\"stroke:rgb(43.9%,50.2%,56.5%);stroke-opacity:1.0;stroke-width:1.0\"></line><text x=\"0\" y=\"6\" style=\"alignment-baseline:hanging;fill:rgb(43.9%,50.2%,56.5%);fill-opacity:1.0;font-size:10px;font-weight:normal;stroke:none;text-anchor:middle\"></text></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "modules[\"toyplot/tables\"] = (function()\n",
       "    {\n",
       "        var tables = [];\n",
       "\n",
       "        var module = {};\n",
       "\n",
       "        module.set = function(owner, key, names, columns)\n",
       "        {\n",
       "            tables.push({owner: owner, key: key, names: names, columns: columns});\n",
       "        }\n",
       "\n",
       "        module.get = function(owner, key)\n",
       "        {\n",
       "            for(var i = 0; i != tables.length; ++i)\n",
       "            {\n",
       "                var table = tables[i];\n",
       "                if(table.owner != owner)\n",
       "                    continue;\n",
       "                if(table.key != key)\n",
       "                    continue;\n",
       "                return {names: table.names, columns: table.columns};\n",
       "            }\n",
       "        }\n",
       "\n",
       "        module.get_csv = function(owner, key)\n",
       "        {\n",
       "            var table = module.get(owner, key);\n",
       "            if(table != undefined)\n",
       "            {\n",
       "                var csv = \"\";\n",
       "                csv += table.names.join(\",\") + \"\\n\";\n",
       "                for(var i = 0; i != table.columns[0].length; ++i)\n",
       "                {\n",
       "                  for(var j = 0; j != table.columns.length; ++j)\n",
       "                  {\n",
       "                    if(j)\n",
       "                      csv += \",\";\n",
       "                    csv += table.columns[j][i];\n",
       "                  }\n",
       "                  csv += \"\\n\";\n",
       "                }\n",
       "                return csv;\n",
       "            }\n",
       "        }\n",
       "\n",
       "        return module;\n",
       "    })();\n",
       "modules[\"toyplot/root/id\"] = \"tf61d0355547449b39a655c17631fed6e\";\n",
       "modules[\"toyplot/root\"] = (function(root_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + root_id);\n",
       "    })(modules[\"toyplot/root/id\"]);\n",
       "modules[\"toyplot/canvas/id\"] = \"teca4e46dea6c433b84a67bab9b740d9e\";\n",
       "modules[\"toyplot/canvas\"] = (function(canvas_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + canvas_id);\n",
       "    })(modules[\"toyplot/canvas/id\"]);\n",
       "modules[\"toyplot/menus/context\"] = (function(root, canvas)\n",
       "    {\n",
       "        var wrapper = document.createElement(\"div\");\n",
       "        wrapper.innerHTML = \"<ul class='toyplot-context-menu' style='background:#eee; border:1px solid #b8b8b8; border-radius:5px; box-shadow: 0px 0px 8px rgba(0%,0%,0%,0.25); margin:0; padding:3px 0; position:fixed; visibility:hidden;'></ul>\"\n",
       "        var menu = wrapper.firstChild;\n",
       "\n",
       "        root.appendChild(menu);\n",
       "\n",
       "        var items = [];\n",
       "\n",
       "        var ignore_mouseup = null;\n",
       "        function open_menu(e)\n",
       "        {\n",
       "            var show_menu = false;\n",
       "            for(var index=0; index != items.length; ++index)\n",
       "            {\n",
       "                var item = items[index];\n",
       "                if(item.show(e))\n",
       "                {\n",
       "                    item.item.style.display = \"block\";\n",
       "                    show_menu = true;\n",
       "                }\n",
       "                else\n",
       "                {\n",
       "                    item.item.style.display = \"none\";\n",
       "                }\n",
       "            }\n",
       "\n",
       "            if(show_menu)\n",
       "            {\n",
       "                ignore_mouseup = true;\n",
       "                menu.style.left = (e.clientX + 1) + \"px\";\n",
       "                menu.style.top = (e.clientY - 5) + \"px\";\n",
       "                menu.style.visibility = \"visible\";\n",
       "                e.stopPropagation();\n",
       "                e.preventDefault();\n",
       "            }\n",
       "        }\n",
       "\n",
       "        function close_menu()\n",
       "        {\n",
       "            menu.style.visibility = \"hidden\";\n",
       "        }\n",
       "\n",
       "        function contextmenu(e)\n",
       "        {\n",
       "            open_menu(e);\n",
       "        }\n",
       "\n",
       "        function mousemove(e)\n",
       "        {\n",
       "            ignore_mouseup = false;\n",
       "        }\n",
       "\n",
       "        function mouseup(e)\n",
       "        {\n",
       "            if(ignore_mouseup)\n",
       "            {\n",
       "                ignore_mouseup = false;\n",
       "                return;\n",
       "            }\n",
       "            close_menu();\n",
       "        }\n",
       "\n",
       "        function keydown(e)\n",
       "        {\n",
       "            if(e.key == \"Escape\" || e.key == \"Esc\" || e.keyCode == 27)\n",
       "            {\n",
       "                close_menu();\n",
       "            }\n",
       "        }\n",
       "\n",
       "        canvas.addEventListener(\"contextmenu\", contextmenu);\n",
       "        canvas.addEventListener(\"mousemove\", mousemove);\n",
       "        document.addEventListener(\"mouseup\", mouseup);\n",
       "        document.addEventListener(\"keydown\", keydown);\n",
       "\n",
       "        var module = {};\n",
       "        module.add_item = function(label, show, activate)\n",
       "        {\n",
       "            var wrapper = document.createElement(\"div\");\n",
       "            wrapper.innerHTML = \"<li class='toyplot-context-menu-item' style='background:#eee; color:#333; padding:2px 20px; list-style:none; margin:0; text-align:left;'>\" + label + \"</li>\"\n",
       "            var item = wrapper.firstChild;\n",
       "\n",
       "            items.push({item: item, show: show});\n",
       "\n",
       "            function mouseover()\n",
       "            {\n",
       "                this.style.background = \"steelblue\";\n",
       "                this.style.color = \"white\";\n",
       "            }\n",
       "\n",
       "            function mouseout()\n",
       "            {\n",
       "                this.style.background = \"#eee\";\n",
       "                this.style.color = \"#333\";\n",
       "            }\n",
       "\n",
       "            function choose_item(e)\n",
       "            {\n",
       "                close_menu();\n",
       "                activate();\n",
       "\n",
       "                e.stopPropagation();\n",
       "                e.preventDefault();\n",
       "            }\n",
       "\n",
       "            item.addEventListener(\"mouseover\", mouseover);\n",
       "            item.addEventListener(\"mouseout\", mouseout);\n",
       "            item.addEventListener(\"mouseup\", choose_item);\n",
       "            item.addEventListener(\"contextmenu\", choose_item);\n",
       "\n",
       "            menu.appendChild(item);\n",
       "        };\n",
       "        return module;\n",
       "    })(modules[\"toyplot/root\"],modules[\"toyplot/canvas\"]);\n",
       "modules[\"toyplot/io\"] = (function()\n",
       "    {\n",
       "        var module = {};\n",
       "        module.save_file = function(mime_type, charset, data, filename)\n",
       "        {\n",
       "            var uri = \"data:\" + mime_type + \";charset=\" + charset + \",\" + data;\n",
       "            uri = encodeURI(uri);\n",
       "\n",
       "            var link = document.createElement(\"a\");\n",
       "            if(typeof link.download != \"undefined\")\n",
       "            {\n",
       "              link.href = uri;\n",
       "              link.style = \"visibility:hidden\";\n",
       "              link.download = filename;\n",
       "\n",
       "              document.body.appendChild(link);\n",
       "              link.click();\n",
       "              document.body.removeChild(link);\n",
       "            }\n",
       "            else\n",
       "            {\n",
       "              window.open(uri);\n",
       "            }\n",
       "        };\n",
       "        return module;\n",
       "    })();\n",
       "modules[\"toyplot.coordinates.Axis\"] = (\n",
       "        function(canvas)\n",
       "        {\n",
       "            function sign(x)\n",
       "            {\n",
       "                return x < 0 ? -1 : x > 0 ? 1 : 0;\n",
       "            }\n",
       "\n",
       "            function mix(a, b, amount)\n",
       "            {\n",
       "                return ((1.0 - amount) * a) + (amount * b);\n",
       "            }\n",
       "\n",
       "            function log(x, base)\n",
       "            {\n",
       "                return Math.log(Math.abs(x)) / Math.log(base);\n",
       "            }\n",
       "\n",
       "            function in_range(a, x, b)\n",
       "            {\n",
       "                var left = Math.min(a, b);\n",
       "                var right = Math.max(a, b);\n",
       "                return left <= x && x <= right;\n",
       "            }\n",
       "\n",
       "            function inside(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.min, range, segment.range.max))\n",
       "                        return true;\n",
       "                }\n",
       "                return false;\n",
       "            }\n",
       "\n",
       "            function to_domain(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.bounds.min, range, segment.range.bounds.max))\n",
       "                    {\n",
       "                        if(segment.scale == \"linear\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            return mix(segment.domain.min, segment.domain.max, amount)\n",
       "                        }\n",
       "                        else if(segment.scale[0] == \"log\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            var base = segment.scale[1];\n",
       "                            return sign(segment.domain.min) * Math.pow(base, mix(log(segment.domain.min, base), log(segment.domain.max, base), amount));\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            var axes = {};\n",
       "\n",
       "            function display_coordinates(e)\n",
       "            {\n",
       "                var current = canvas.createSVGPoint();\n",
       "                current.x = e.clientX;\n",
       "                current.y = e.clientY;\n",
       "\n",
       "                for(var axis_id in axes)\n",
       "                {\n",
       "                    var axis = document.querySelector(\"#\" + axis_id);\n",
       "                    var coordinates = axis.querySelector(\".toyplot-coordinates-Axis-coordinates\");\n",
       "                    if(coordinates)\n",
       "                    {\n",
       "                        var projection = axes[axis_id];\n",
       "                        var local = current.matrixTransform(axis.getScreenCTM().inverse());\n",
       "                        if(inside(local.x, projection))\n",
       "                        {\n",
       "                            var domain = to_domain(local.x, projection);\n",
       "                            coordinates.style.visibility = \"visible\";\n",
       "                            coordinates.setAttribute(\"transform\", \"translate(\" + local.x + \")\");\n",
       "                            var text = coordinates.querySelector(\"text\");\n",
       "                            text.textContent = domain.toFixed(2);\n",
       "                        }\n",
       "                        else\n",
       "                        {\n",
       "                            coordinates.style.visibility= \"hidden\";\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            canvas.addEventListener(\"click\", display_coordinates);\n",
       "\n",
       "            var module = {};\n",
       "            module.show_coordinates = function(axis_id, projection)\n",
       "            {\n",
       "                axes[axis_id] = projection;\n",
       "            }\n",
       "\n",
       "            return module;\n",
       "        })(modules[\"toyplot/canvas\"]);\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t14d72f9fec834daf987af88a098a04e3\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[0.0, 2000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t93f3ceba288a42638a3724daead8545b\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[2001.0, 4000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"tbabcb03ce25d463299d6f2a6189b6e8b\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[4001.0, 6000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"ta2b528c2ef56434793d4f40203a283fe\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[6001.0, 8000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"tbd21d96938ce49939647154b2e7665d2\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[8001.0, 10000.0], [0.0, 0.0], [1.0, 1.0]],\"toyplot\");\n",
       "(function(axis, axis_id, projection)\n",
       "        {\n",
       "            axis.show_coordinates(axis_id, projection);\n",
       "        })(modules[\"toyplot.coordinates.Axis\"],\"t76d36694ce7f432eb5465e0f6a8f47f4\",[{\"domain\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 10000.0, \"min\": 0.0}, \"range\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 200.0, \"min\": 0.0}, \"scale\": \"linear\"}]);\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"ta97350ec521748418ca035b28844511b\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[0.0, 13.0], [6889.938482828021, 6889.938482828021], [7177.019252945855, 7177.019252945855]],\"toyplot\");\n",
       "(function(tables, context_menu, io, owner_id, key, label, names, columns, filename)\n",
       "        {\n",
       "            tables.set(owner_id, key, names, columns);\n",
       "\n",
       "            var owner = document.querySelector(\"#\" + owner_id);\n",
       "            function show_item(e)\n",
       "            {\n",
       "                return owner.contains(e.target);\n",
       "            }\n",
       "\n",
       "            function choose_item()\n",
       "            {\n",
       "                io.save_file(\"text/csv\", \"utf-8\", tables.get_csv(owner_id, key), filename + \".csv\");\n",
       "            }\n",
       "\n",
       "            context_menu.add_item(\"Save \" + label + \" as CSV\", show_item, choose_item);\n",
       "        })(modules[\"toyplot/tables\"],modules[\"toyplot/menus/context\"],modules[\"toyplot/io\"],\"t00aede1bbebf45ed86311f7e42a02423\",\"data\",\"fill data\",[\"x\", \"y0\", \"y1\"],[[0.0, 1.0, 1.0, 12.0, 12.0, 13.0], [6889.938482828021, 5741.615402356684, 0.0, 0.0, 5741.615402356684, 6889.938482828021], [6889.938482828021, 6889.938482828021, 6889.938482828021, 6889.938482828021, 6889.938482828021, 6889.938482828021]],\"toyplot\");\n",
       "(function(axis, axis_id, projection)\n",
       "        {\n",
       "            axis.show_coordinates(axis_id, projection);\n",
       "        })(modules[\"toyplot.coordinates.Axis\"],\"tfdc3c3b2c16147eaaedbd49cb9fcb095\",[{\"domain\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 7177.019252945855, \"min\": -1827.3995930591586}, \"range\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 275.0, \"min\": 0.0}, \"scale\": \"linear\"}]);\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "postsim.draw_tree_sequence();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NOTE:** The *default* Wright Fisher model in shadie is the classic model, which has no selection. Even if non-neutral mutation types are defined and occur, they do not affect the simulation outcome. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Breaking the Assumptions of Wright-Fisher\n",
    "There is a parameter in the `shadie` Wright-Fisher models, `selection`, that controls what kind of selection can occur in the model. Adding selection will break the **neutrality** assumption of the Wright-Fisher.\n",
    "\n",
    "There are two options for adding selection: \n",
    "- soft selection: fitness affects mating success only\n",
    "- hard selection: fitness affects survival only\n",
    "\n",
    "Soft and hard selection are implemented in the examples below. These examples also break the **no recombination** assumption, simply by passing a non-zero value to the `recomb_rate` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as WF_soft_model:\n",
    "    WF_soft_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        recomb_rate=1e-9,\n",
    "        file_out=\"WF_soft.trees\",\n",
    "    )\n",
    "    WF_soft_model.reproduction.wright_fisher(\n",
    "        pop_size=1000,\n",
    "        selection= \"soft\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as WF_hard_model:\n",
    "    WF_hard_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        recomb_rate=1e-9,\n",
    "        file_out=\"WF_hard.trees\",\n",
    "    )\n",
    "    WF_hard_model.reproduction.wright_fisher(\n",
    "        pop_size=1000,\n",
    "        selection= \"hard\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WF models incorporating selection may be more appropriate to use as control simulations to compare against simulations with complex life cycle models. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alternative Models in `shadie` based on Wright-Fisher \n",
    "Because the life cycle models in `shadie` incorporate alternations of generations and both haploid and diploid individuals, there are further modifications to the Wright-Fisher model that may be useful. The same three options for selection (\"none\", \"soft\" and \"hard) are available for each of them.\n",
    "### Moran Model\n",
    "The Moran model is a simple alternative to Wright-Fisher that allows overlapping generations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as moran_model:\n",
    "    moran_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        recomb_rate= 0.0,\n",
    "        file_out=\"moran.trees\",\n",
    "    )\n",
    "    moran_model.reproduction.moran(pop_size=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Haploid Models\n",
    "The default Wright-Fisher model usually assumes diploid individuals. There are two haploid Wright-Fisher models in `shadie`:\n",
    "- clonal: A single parent is chosen N times with replacement; each time a parents is chosen it produces one offspring\n",
    "- sexual: The parent genomes undergo reocmbination and one offspring is produced per pair. This requires recombination rate != 0.0\n",
    "Both models have all three selection options: \"none\", \"soft\", and \"hard\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as WF_1N_clonal_model:\n",
    "    WF_1N_clonal_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        file_out=\"1N_clonal.trees\",\n",
    "    )\n",
    "    WF_1N_clonal_model.reproduction.wright_fisher_haploid_clonal(pop_size=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as WF_1N_sexual_model:\n",
    "    WF_1N_sexual_model.initialize(\n",
    "        chromosome=default_chrom,\n",
    "        recomb_rate=1e-9,\n",
    "        file_out=\"1N_sexual.trees\",\n",
    "    )\n",
    "    WF_1N_sexual_model.reproduction.wright_fisher_haploid_sexual(\n",
    "        pop_size=1000,\n",
    "        selection=\"soft\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternation-of-Generations \"Wright-Fisher\"\n",
    "This is a highly altered Wright-Fisher model that incorporates alternation of generations, which may be useful as a control or comparison to more complex models with alternation-of-generations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with shadie.Model() as WF_altgen_model:\n",
    "    WF_altgen_model.initialize(\n",
    "        chromosome=default_chrom, \n",
    "        recomb_rate= 0.0,\n",
    "        file_out=\"moran.trees\",\n",
    "    )\n",
    "    WF_altgen_model.reproduction.wright_fisher_altgen(\n",
    "        spo_pop_size=1000,\n",
    "        gam_pop_size=1000,\n",
    "        selection=\"hard\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "initialize() {\n",
      "\n",
      "  // model type\n",
      "  initializeSLiMModelType(\"nonWF\");\n",
      "\n",
      "  // config\n",
      "  initializeRecombinationRate(0.0, 10000);\n",
      "  initializeMutationRate(1e-08);\n",
      "  initializeTreeSeq(simplificationInterval=NULL);\n",
      "\n",
      "  // MutationType init\n",
      "  initializeMutationType('m2', 0.1, 'g', -3.0, 1.5);\n",
      "  initializeMutationType('m3', 0.8, 'e', 0.04);\n",
      "  c(m3, m2).haploidDominanceCoeff = 1.0;\n",
      "  c(m3, m2).convertToSubstitution = T;\n",
      "\n",
      "  \n",
      "\n",
      "  // ElementType init\n",
      "  initializeGenomicElementType('g1', c(m2,m3), c(8,0.1));\n",
      "  initializeGenomicElementType('g2', m2, 1);\n",
      "\n",
      "  // Chromosome (GenomicElement init)\n",
      "  types = rep(g1, asInteger(floor(1999/3)));\n",
      "  starts = 2001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "  types = rep(g2, asInteger(floor(1999/3)));\n",
      "  starts = 4001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "  types = rep(g1, asInteger(floor(1999/3)));\n",
      "  starts = 6001 + seqLen(integerDiv(1999, 3)) * 3;\n",
      "  ends = starts + 1;\n",
      "  initializeGenomicElement(types, starts, ends);\n",
      "\n",
      "\n",
      "  // constants (Population sizes and others)\n",
      "  defineConstant('SPO_POP_SIZE', 1000);\n",
      "  defineConstant('GAM_POP_SIZE', 1000);\n",
      "\n",
      "  // globals (metadata dictionary)\n",
      "  defineGlobal('METADATA', Dictionary('model', 'shadie haploid clonal WF', 'length', '1000', 'spo_pop_size', '1000', 'gam_pop_size', '1000', 'spo_mutation_rate', '1e-08', 'selection', 'hard', 'recombination_rate', '0.0'));\n",
      "\n",
      "  // extra scripts (Optional)\n",
      "  \n",
      "}\n",
      "\n",
      "\n",
      "// executes before offspring are generated\n",
      "// define starting haploid population.\n",
      "\n",
      "1 first() {\n",
      "    sim.addSubpop('p1', SPO_POP_SIZE);\n",
      "sim.addSubpop('p0', 0);\n",
      "}\n",
      "\n",
      "\n",
      "// generates offspring\n",
      "// clonal random mating; mating success weighted by fitness.\n",
      "reproduction(p1 ) {\n",
      "    // parents are chosen randomly\n",
      "    fitness = p1.cachedFitness(NULL);\n",
      "    parents = sample(p1.individuals, GAM_POP_SIZE, replace=T);\n",
      "    for (i in seqLen(GAM_POP_SIZE)){\n",
      "        breaks = sim.chromosome.drawBreakpoints(parents[i]);\n",
      "        p0.addRecombinant(parents.genome1[i], parents.genome2[i], breaks, NULL, NULL, NULL);\n",
      "    }\n",
      "    self.active = 0;\n",
      "}\n",
      "\n",
      "\n",
      "// generates offspring\n",
      "// clonal random mating; mating success weighted by fitness.\n",
      "reproduction(p0 ) {\n",
      "    // parents are chosen randomly\n",
      "    fitness = p0.cachedFitness(NULL);\n",
      "    parents1 = sample(p0.individuals, SPO_POP_SIZE, replace=T);\n",
      "    parents2 = sample(p0.individuals, SPO_POP_SIZE, replace=T);\n",
      "    for (i in seqLen(SPO_POP_SIZE))\n",
      "        p1.addRecombinant(parents1.genome1[i], NULL, NULL, parents2.genome1[i], NULL, NULL);\n",
      "    self.active = 0;\n",
      "}\n",
      "\n",
      "\n",
      "// executes after offspring are generated\n",
      "// Fitness scaling for hard selection\n",
      "\n",
      "early() {\n",
      "    // diploids (p1) just generated haploid into p0\n",
      "    if (community.tick % 2 == 0) {\n",
      "        // fitness affects haploid survival\n",
      "        p0.fitnessScaling = GAM_POP_SIZE / p0.individualCount;\n",
      "        }\n",
      "    // haploids (p0) just generated diploids into p1\n",
      "    else {\n",
      "        //fitness affects diploid survival\n",
      "        p1.fitnessScaling = SPO_POP_SIZE / p1.individualCount;\n",
      "    }\n",
      "}\n",
      "\n",
      "\n",
      "// implements survival adjustments\n",
      "// non-overlapping generations\n",
      "survival() {\n",
      "    return (individual.age == 0);\n",
      "}\n",
      "\n",
      "\n",
      "// executes after selection occurs\n",
      "// end of sim; save .trees file\n",
      "\n",
      "2001 late() {\n",
      "    sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);\n",
      "    sim.treeSeqOutput('moran.trees', metadata = METADATA);\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(WF_altgen_model.script)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
